dallinger/recruiters.py

```diff
@@ -564,69 +599,89 @@ def parse_spec(self):
             recruiters.append((name, count))
         return recruiters
 
-    def pick_recruiter(self):
-        """Pick the next recruiter to use.
+    def recruiters(self, n=1):
+        """Iterator that provides recruiters along with the participant
+        count to be recruited for up to `n` participants.
 
         We use the `Recruitment` table in the db to keep track of
         how many recruitments have been requested using each recruiter.
         We'll use the first one from the specification that
         hasn't already reached its quota.
         """
-        counts = dict(
-            session.query(
-                Recruitment.recruiter_id,
-                func.count(Recruitment.id)
-            ).group_by(Recruitment.recruiter_id).all()
-        )
-
-        for recruiter_id, target_count in self.spec:
-            count = counts.get(recruiter_id, 0)
-            if count >= target_count:
-                # This recruiter quota was reached;
-                # move on to the next one.
-                counts[recruiter_id] = count - target_count
-                continue
-            else:
-                # Quota is still available; let's use it.
-                break
-        else:
-            raise Exception(
-                'Reached quota for all recruiters. '
-                'Not sure which one to use now.'
+        recruit_count = 0
+        while recruit_count <= n:
+            counts = dict(
+                session.query(
+                    Recruitment.recruiter_id,
+                    func.count(Recruitment.id)
+                ).group_by(Recruitment.recruiter_id).all()
             )
+            for recruiter_id, target_count in self.spec:
+                remaining = 0
+                count = counts.get(recruiter_id, 0)
+                if count >= target_count:
+                    # This recruiter quota was reached;
+                    # move on to the next one.
+                    counts[recruiter_id] = count - target_count
+                    continue
+                else:
+                    # Quota is still available; let's use it.
+                    remaining = target_count - count
+                    break
+            else:
+                raise StopIteration
```

Comment:
I believe `StopIteration` is considered an implementation detail of generators and that a bare `return` is the preferred way of terminating the loop.

Reply:
I seem to recall having a `return` in an iterator being a syntax error or similar, but that was probably back in the Python 2.2 days. I'll fix this.

Reply:
It's only a syntax error if there is a return value, a bare return is allowed. Crazy, huh?

Reply:
I've been raising `StopIteration` all these years for nothing! Thanks.